R/utilities-dimensions.R

```diff
@@ -0,0 +1,162 @@
+}
+
+#' Calculate factor for converting values of one dimension into another
+#'
+#' @param dimension1 String name of the first dimension
+#' @param dimension2 String name fo the second dimension
+#' @param mw Double value of the molecular weight of the molecule. Used
+#' for conversion between molar and mass concentrations or amounts. Can be \code{NULL}
+#' for other dimensions (default)
+#'
+#' @return A numerical factor the values of the first dimensions must be multiplied by to
+#' be converted to second dimension. The conversion assumes values being in base units
+#' of both dimensions.
+#' If the values cannot be transformed between the dimensions, an error is thrown.
+#' @export
+dimensionsConversionFactor <- function(dimension1, dimension2, mw = NULL) {
+  # No conversion if the dimensions are equall
+  validateIsString(c(dimension1, dimension2))
+  if (dimension1 == dimension2) {
+    return(1)
+  }
+
+  # Dimensionless <-> Fraction
+  if (dimension1 == Dimensions$Dimensionless) {
+    if (dimension2 == Dimensions$Fraction) {
+      return(1)
+    }
+  }
+  if (dimension1 == Dimensions$Fraction) {
+    if (dimension2 == Dimensions$Dimensionless) {
+      return(1)
+    }
+  }
+
+  # Concentration (molar) <-> Concentration (mass)
+  if (dimension1 == Dimensions$`Concentration (molar)`) {
+    if (dimension2 == Dimensions$`Concentration (mass)`) {
+      if (is.null(mw)) {
+        stop(messages$errorCannotConvertDimensionsNoMW(dimension1, dimension2))
+      }
+      return(1 * 1e-6 * mw * 1e-3)
+    }
+  }
+  if (dimension1 == Dimensions$`Concentration (mass)`) {
+    if (dimension2 == Dimensions$`Concentration (molar)`) {
+      if (is.null(mw)) {
+        stop(messages$errorCannotConvertDimensionsNoMW(dimension1, dimension2))
+      }
+      return(1 * 1e3 / mw * 1e6)
+    }
+  }
+
+  # Amount (molar) <-> Mass
+  if (dimension1 == Dimensions$Amount) {
+    if (dimension2 == Dimensions$Mass) {
+      if (is.null(mw)) {
+        stop(messages$errorCannotConvertDimensionsNoMW(dimension1, dimension2))
+      }
+      return(1 * 1e-6 * mw * 1e-3)
+    }
+  }
+  if (dimension1 == Dimensions$Mass) {
+    if (dimension2 == Dimensions$Amount) {
+      if (is.null(mw)) {
+        stop(messages$errorCannotConvertDimensionsNoMW(dimension1, dimension2))
+      }
+      return(1 * 1e3 / mw * 1e6)
+    }
+  }
+
+  stop(messages$errorCannotConvertDimensions(dimension1, dimension2))
+}
+
+#' Get conversion factor from one unit to another
+#'
+#' @param fromUnit String name of the unit to convert from
+#' @param toUnit String name of the unit to convert to
+#' @param dimension Dimension of the units
+#'
+#' @return Numerical value. A value in \code{fromUnit} multiplied  by the factor
+#' corresponds to the value in \code{toUnit}
+#' @export
+getUnitConversionFactor <- function(fromUnit, toUnit, dimension) {
+  validateDimension(dimension)
+  validateUnit(fromUnit, dimension)
+  validateUnit(toUnit, dimension)
+
+  dimensionTask <- getDimensionTask()

```

Comment:
I am not sure why you need sthg like this ? toUnit converts already everything from you. .The goal is to abstract exactly all of this away

Reply:
The idea is to use the conversion capabilities of OSPS without having the simulation objects (quantities). E.g. read the data from excel and plot it in another unit. `toUnit` requires a Quantity.

Reply:
nope

```
toUnit <- function(quantityOrDimension, values, targetUnit, molWeight = NULL) {
```

Specifically the parameter is called quantityOrDimension.

None of the proposed changes regarding reimplementing conversion by hand belong in Core. Helper function to ensure that dimension and unit etc.. exists, yes.  The rest probably not :)